geopandas/sindex.py

```diff
@@ -70,6 +91,141 @@ def __init__(self, geometry):
+            predicate : {None, 'intersects', 'within', 'contains', 'overlaps', 'crosses', 'touches'}, optional
+                If predicate is provided, a prepared version of the input geometry is tested using
+                the predicate function against each item in the index whose extent intersects the
+                envelope of the input geometry: predicate(geometry, tree_geometry).
+            sort : bool, default False
+                If True, the results will be sorted in ascending order. If False, results are
+                often sorted but there is no guarantee.
+
+            Returns
+            -------
+            matches : ndarray of shape (n_results, )
+                Integer indices for matching geometries from the spatial index.
+            """  # noqa: E501
+            # handle invalid predicates
+            if predicate not in self.valid_query_predicates:
+                raise ValueError(
+                    "Got `predicate` = `{}`, `predicate` must be one of {}".format(
+                        predicate, self.valid_query_predicates
+                    )
+                )
+            # handle empty / invalid geometries
+            if geometry is None:
+                # this is the behavior in pygeos.strtree.query, we mimic it here
+                return np.array([], dtype=np.intp)
+            if not isinstance(geometry, BaseGeometry):
+                raise TypeError(
+                    "Got `geometry` of type `{}`, `geometry` must be ".format(
+                        type(geometry)
+                    )
+                    + "a shapely geometry."
+                )
+            if geometry.is_empty:
+                return np.array([], dtype=np.intp)
+
+            # get bounds
+            bounds = geometry.bounds  # rtree operates on bounds
+            tree_query = list(self.intersection(bounds, objects=False))
+
+            if not tree_query:
+                return np.array([], dtype=np.intp)
+
+            # check predicate
+            if predicate in ("intersects", "within"):
+                # only contains and intersects are supported by
+                # prepared geometries, see note below regarding within
+                res = []
+                if predicate == "within":
+                    # since these are inverse, we can flip the operation
+                    # and test with prepared predicates from tree
+                    predicate = "contains"
+                for i in tree_query:
+                    if self._prepared_geometries[i] is None:
+                        # if not already prepared, prepare and cache
+                        self._prepared_geometries[i] = prep(self._geometries[i])
+                    if getattr(self._prepared_geometries[i], predicate)(geometry):
+                        res.append(i)
+                tree_query = res
+            elif predicate == "contains" and len(tree_query) > 1:
```

Comment:
Limiting this case to `len(tree_query) > 1` is not immediately obvious.  Am I correct in thinking that even if there is 1 hit from the tree, it does not guarantee that the input geometry `contains` the tree geometry?  As in, their bounding boxes intersect, but without evaluating the predicate, we don't know how the actual geometries relate to each other.

Reply:
This is a good catch. The way it is written right now, it's a bug. I guess my tests are not very good (I'll add one for this specific case). I'm going to rewrite as follows:

```python3
elif predicate == "contains":
    if len(tree_idx) > MIN_MATCHES_FOR_PREP:
        # prepare this geometry
        geometry = prep(geometry)
     tree_idx = [
         i
         for i in tree_idx
         if getattr(geometry, predicate)(self._geometries[i])
     ]
```

As an aside, what I just called `MIN_MATCHES_FOR_PREP` above is currently just `1`. Do you think this is a check we should do? Is there a better number? Or should we always prep?

Reply:
@jorisvandenbossche answered below. We are going to always use prepared geometries.